File: articles/food-chemistry-molecular-basis-nutrition/advanced/go/food_chemistry_metrics.go
// Food Chemistry and the Molecular Basis of Nutrition
// Go food chemistry analytics example.
// Synthetic educational code only.

package main

import (
	"fmt"
	"math"
	"os"
)

type Food struct {
	Name                  string
	EnergyKcal            float64
	ProteinG              float64
	FiberG                float64
	PotassiumMg           float64
	ProcessingIntensity   float64
	RetentionFactor       float64
	VitaminCMg             float64
	BioavailabilityFactor  float64
	IronMg                float64
}

func clamp01(x float64) float64 {
	return math.Max(0.0, math.Min(1.0, x))
}

func nutrientDensity(food Food) float64 {
	beneficial := 0.40*clamp01(food.ProteinG/25.0) +
		0.35*clamp01(food.FiberG/12.0) +
		0.25*clamp01(food.PotassiumMg/800.0)

	energyFactor := math.Max(food.EnergyKcal/100.0, 0.5)
	return beneficial / energyFactor
}

func retainedVitaminC(food Food) float64 {
	return food.VitaminCMg * food.RetentionFactor
}

func bioavailableIron(food Food) float64 {
	return food.IronMg * food.RetentionFactor * food.BioavailabilityFactor
}

func main() {
	os.MkdirAll("../outputs/tables", 0755)

	foods := []Food{
		{"lentils_cooked", 230, 18, 15, 730, 0.25, 0.88, 3, 0.16, 6.6},
		{"orange_segments", 80, 1.5, 4, 240, 0.15, 0.82, 70, 0.08, 0.2},
		{"spinach_cooked", 45, 5, 4, 840, 0.35, 0.72, 18, 0.07, 6.4},
	}

	file, err := os.Create("../outputs/tables/go_food_chemistry_metrics.csv")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	fmt.Fprintln(file, "food,nutrient_density,retained_vitamin_c_mg,bioavailable_iron_mg")

	for _, food := range foods {
		fmt.Fprintf(
			file,
			"%s,%.6f,%.6f,%.6f\n",
			food.Name,
			nutrientDensity(food),
			retainedVitaminC(food),
			bioavailableIron(food),
		)
	}

	fmt.Println("Go food chemistry metrics complete.")
}

File: articles/food-chemistry-molecular-basis-nutrition/advanced/rust/food_chemistry_metrics.rs
// Food Chemistry and the Molecular Basis of Nutrition
// Rust food chemistry scoring example.
// Synthetic educational code only.

use std::fs::{create_dir_all, File};
use std::io::{Result, Write};

#[derive(Debug)]
struct Food {
    name: &'static str,
    energy_kcal: f64,
    protein_g: f64,
    fiber_g: f64,
    potassium_mg: f64,
    starch_g: f64,
    sugars_g: f64,
    particle_accessibility: f64,
    processing_intensity: f64,
}

fn clamp01(x: f64) -> f64 {
    x.max(0.0).min(1.0)
}

fn nutrient_density(food: &Food) -> f64 {
    let beneficial =
        0.40 * clamp01(food.protein_g / 25.0) +
        0.35 * clamp01(food.fiber_g / 12.0) +
        0.25 * clamp01(food.potassium_mg / 800.0);

    let energy_factor = (food.energy_kcal / 100.0).max(0.5);
    beneficial / energy_factor
}

fn glycemic_accessibility(food: &Food) -> f64 {
    let fiber_protection = clamp01(food.fiber_g / 12.0);

    clamp01(
        0.32 * clamp01(food.starch_g / 30.0) +
        0.24 * clamp01(food.sugars_g / 20.0) +
        0.22 * food.particle_accessibility +
        0.16 * food.processing_intensity -
        0.20 * fiber_protection
    )
}

fn main() -> Result<()> {
    create_dir_all("../outputs/tables").ok();

    let foods = vec![
        Food { name: "lentils_cooked", energy_kcal: 230.0, protein_g: 18.0, fiber_g: 15.0, potassium_mg: 730.0, starch_g: 22.0, sugars_g: 3.0, particle_accessibility: 0.45, processing_intensity: 0.25 },
        Food { name: "white_bread", energy_kcal: 160.0, protein_g: 5.0, fiber_g: 1.0, potassium_mg: 70.0, starch_g: 24.0, sugars_g: 3.0, particle_accessibility: 0.88, processing_intensity: 0.75 },
        Food { name: "oats_cooked", energy_kcal: 180.0, protein_g: 10.0, fiber_g: 8.0, potassium_mg: 180.0, starch_g: 18.0, sugars_g: 1.0, particle_accessibility: 0.50, processing_intensity: 0.30 },
    ];

    let mut file = File::create("../outputs/tables/rust_food_chemistry_metrics.csv")?;
    writeln!(file, "food,nutrient_density,glycemic_accessibility")?;

    for food in &foods {
        writeln!(
            file,
            "{},{:.6},{:.6}",
            food.name,
            nutrient_density(food),
            glycemic_accessibility(food)
        )?;
    }

    println!("Rust food chemistry metrics complete.");
    Ok(())
}
